src/goto-programs/validate_goto_model.cpp

```diff
@@ -0,0 +1,223 @@
+/*******************************************************************\
+Module: Validate Goto Programs
+
+Author: Diffblue
+
+Date: Oct 2018
+
+\*******************************************************************/
+
+#include "validate_goto_model.h"
+
+#include <set>
+
+#include <util/invariant.h>
+
+#include "goto_functions.h"
+
+namespace
+{
+class validate_goto_modelt
+{
+public:
+  using function_mapt = goto_functionst::function_mapt;
+
+  validate_goto_modelt(
+    const goto_functionst &goto_functions,
+    const validation_modet vm)
+    : vm{vm}, function_map{goto_functions.function_map}
+  {
+  }
+
+  void
+  do_goto_program_checks(goto_model_validation_optionst validation_options);
+
+private:
+  /// Check the goto_program has an entry point
+  ///
+  /// NB goto-cc -c will generate goto-programs without entry-points
+  /// until they are linked.
+  void entry_point_exists();
+
+  /// Check that no function calls via function pointer are present
+  void function_pointer_calls_removed();
+
+  /// Check returns have been removed
+  ///
+  /// NB Calls via function pointer must have been removed already when
+  /// removing returns, thus enabling this check also enables the check
+  /// that all calls via function pointer have been removed
+  ///
+  /// Sub-checks are:
+  /// - no return statements in any of the functions
+  /// - lhs of every \ref code_function_callt instruction is nil
+  /// - all return types are void (of both calls and functions themselves)
+  void check_returns_removed();
+
+  /// Check that for all:
+  /// -# functions that are called or
+  /// -# functions of which the address is taken
+  /// .
+  /// a corresponding entry exists in the function_map
+  ///
+  /// n.b. functions that are only declared and not used will not appear in the
```

Comment:
Super :pick: again: If we do use the nb, I'd prefer if it was consistently spelled as either `n.b.` or `NB`, not both

Reply:
Addressed as above.